### PR TITLE
Filesystem Create/Mount/Umount for thin devices

### PR DESCRIPTION
**Pull request by trgill**
Add Create/Mount/Umount for XFS on thindev
Added to the existing thinpool test:
    Creates a FS
    Mounts FS on temp dir
    Creates a directory on the FS and writes 100 files with small amount of content
    Unmounts 
    Tears down the DM framework
Minor test framework clean up

### CODE REVIEW

File: tests/util/blockdev_utils.rs [hunk 1]
```diff
@@ -17,10 +17,28 @@ use std::path::PathBuf;
 use std::thread;
 use std::time::Duration;
 
+use util::blockdev_utils::tempdir::TempDir;
 use util::test_result::TestError;
 use util::test_result::TestErrorEnum;
 use util::test_result::TestResult;
 
+
```
Comment (mulkieran) [inline tests/util/blockdev_utils.rs]: These seem way too generic to put in blockdev_utils.rs. A new file might not be a bad idea.

File: tests/util/blockdev_utils.rs [hunk 2]
```diff
@@ -17,10 +17,28 @@ use std::path::PathBuf;
...
+        Ok(val) => info!("{:?}", val),
+        Err(err) => {error!("FAILED : {:?}", err);assert!(false);
+        }
+    });
+}
+
+macro_rules! try_log_error {
+    ($e:expr) => (match $e {
+        Ok(val) => info!("{:?}", val),
+        Err(err) =>
```
Comment (mulkieran) [inline tests/util/blockdev_utils.rs]: wierd formatting.

File: src/engine/strat_engine/thindev.rs [hunk 3]
```diff
@@ -77,4 +79,68 @@ impl ThinDev {
...
+        } else {
+            let message = String::from_utf8_lossy(&output.stderr);
+            debug!("stderr: {}", message);
+            return Err(EngineError::Engine(ErrorEnum::Error, message.into()));
+        }
+        Ok(())
+    }
+
+    pub fn mount_fs(&mut self, mount_point: &Path) -> EngineResult<()> {
+        let path_result = self.path().unwrap().unwrap();
```
Comment (trgill) [inline src/engine/strat_engine/thindev.rs]: Yeah, it was a Option wrapped in a Result.  I updated to simplify and added try!

File: tests/util/blockdev_utils.rs (hunk 1 above)
Comment (trgill) [inline tests/util/blockdev_utils.rs]: Ok - moved to test_utils.rs

File: tests/util/blockdev_utils.rs (hunk 2 above)
Comment (trgill) [inline tests/util/blockdev_utils.rs]: fixed - had to wrestle with rustfmt a bit...

File: src/engine/strat_engine/thindev.rs [hunk 4]
```diff
@@ -77,4 +79,68 @@ impl ThinDev {
...
+
+    pub fn path(&self) -> EngineResult<Option<PathBuf>> {
+        match self.dev_info {
+            Some(ref di) => return Ok(di.device().path().clone()),
+            None => return Err(EngineError::Engine(ErrorEnum::Invalid, "No dev_info".into())),
+        }
+    }
+
+    pub fn create_fs(&mut self) -> EngineResult<()> {
+        let path_result = self.path().unwrap().unwrap();
```
Comment (trgill) [inline src/engine/strat_engine/thindev.rs]: ok

**Comment by agrover** [comment]
Hi! Please:
* edit d718f87 commit summary to be <60 chars (just move 2nd sentence down 2 lines?) and fix `LinearDev` spelling
* Combine `ThinDev::setup` and `::new`. Make `new` also do setup. This should let you remove `Option` from all `struct ThinDev` members, which should let member functions be simpler. Possible unused struct members.

**Comment by trgill** [comment]
@agrover I've addressed your comments.

File: src/engine/strat_engine/lineardev.rs [hunk 5]
```diff
@@ -13,22 +13,29 @@ use types::Sectors;
...
 }
 
-/// support use of DM to concatenate blockdevs and creat e a
-/// /dev/mapper/<name> for use as continuous sectors.
+/// Use DM to concatenate a set of blockdevs together into a
+/// /dev/mapper/xxx block device of continuous sectors.
 impl LinearDev {
-    pub fn new(name: &str) -> LinearDev {
-        LinearDev {
+    pub fn new(name: &str, dm: &DM, block_devs: &Vec<&BlockDev>) -> EngineResult<LinearDev> {
```
Comment (mulkieran) [inline src/engine/strat_engine/lineardev.rs]: ```new``` has gotten so ambitious that a comment wouldn't hurt.

File: src/engine/strat_engine/thindev.rs [hunk 6]
```diff
@@ -26,21 +27,32 @@ pub enum ThinStatus {
 
 /// support use of DM to thin provisioned devices over pools
 impl ThinDev {
-    pub fn new(name: &str) -> ThinDev {
-        ThinDev {
+    pub fn new(name: &str,
```
Comment (mulkieran) [inline src/engine/strat_engine/thindev.rs]: header comment here would be helpful, too.

File: src/engine/strat_engine/thindev.rs [hunk 7]
```diff
@@ -54,27 +66,75 @@ impl ThinDev {
...
-        try!(thin_pool.message(dm, &format!("create_thin {}", thin_id)));
-        try!(dm.device_create(&self.name, None, DmFlags::empty()));
-        let table = self.dm_table(&thin_pool.dev_info.unwrap().device().dstr(),
-                                  thin_id,
-                                  length);
-        let id = &DevId::Name(&self.name);
-        self.dev_info = Some(try!(dm.table_load(id, &table)));
-        try!(dm.device_suspend(id, DmFlags::empty()));
+    pub fn create_fs(&mut self) -> EngineResult<()> {
+        let path_result = try!(self.path());
```
Comment (mulkieran) [inline src/engine/strat_engine/thindev.rs]: These two lines might as well be one.

File: src/engine/strat_engine/thindev.rs [hunk 8]
```diff
@@ -54,27 +66,75 @@ impl ThinDev {
...
+            debug!("stderr: {}", message);
+            return Err(EngineError::Engine(ErrorEnum::Error, message.into()));
+        }
         Ok(())
     }
 
-    pub fn teardown(&mut self, dm: &DM) -> EngineResult<()> {
-        try!(dm.device_remove(&DevId::Name(&self.name), DmFlags::empty()));
+    pub fn mount_fs(&mut self, mount_point: &Path) -> EngineResult<()> {
+        let path_result = try!(self.path());
```
Comment (mulkieran) [inline src/engine/strat_engine/thindev.rs]: These two lines could also be merged into one.

File: src/engine/strat_engine/thinpooldev.rs [hunk 9]
```diff
@@ -3,40 +3,60 @@
...
-/// support use of DM to create pools for thin provisioned devices
+/// Use DM to create a "thin-pool".  A "thin-pool" is shared space for
+/// other thin provisioned devices to use.
+///
+/// See section "Setting up a fresh pool device":
+/// https://www.kernel.org/doc/Documentation/device-mapper/thin-provisioning.txt
 impl ThinPoolDev {
-    pub fn new(name: &str) -> ThinPoolDev {
-        ThinPoolDev {
+    pub fn new(name: &str,
```
Comment (mulkieran) [inline src/engine/strat_engine/thinpooldev.rs]: Header comment helpful, comment on impl might be better on ```new``` method.

File: tests/util/blockdev_utils.rs [hunk 10]
```diff
@@ -75,11 +78,24 @@ pub fn get_size(path: &Path) -> TestResult<Sectors> {
...
+pub fn wipe_header(path: &Path) -> TestResult<()> {
     let mut f = try!(OpenOptions::new().write(true).open(path));
     let zeroed = [0u8; SECTOR_SIZE * 16];
 
     try!(f.write_all(&zeroed[..SECTOR_SIZE * 16]));
 
     Ok(())
 }
+
+#[allow(dead_code)]
```
Comment (mulkieran) [inline tests/util/blockdev_utils.rs]: Doest rust still think this is dead?

File: tests/lineardev_tests.rs [hunk 11]
```diff
@@ -75,7 +67,10 @@ fn validate_sizes(name: &str, block_devs: &Vec<&BlockDev>) -> TestResult<()> {
 
     let dm_dev_size = match get_size(path) {
         Ok(s) => s,
-        Err(_) => panic!("Failed to get size for {} ", name),
+        Err(e) => {
```
Comment (mulkieran) [inline tests/lineardev_tests.rs]: I don't know why you don't panic here.

File: src/engine/strat_engine/thindev.rs (hunk 7 above)
Comment (trgill) [inline src/engine/strat_engine/thindev.rs]: If I change this to:

 let dev_path = Path::new(&try!(self.path()));

Am I misunderstanding your comment?
I get "borrowed value does not live long enough: consider using a `let` binding to increase its lifetime temporary value dropped here while still  borrowed"

File: tests/util/blockdev_utils.rs (hunk 10 above)
Comment (trgill) [inline tests/util/blockdev_utils.rs]: Yes.  I'm not sure why - I've tried to figure it out a couple times but never resolved it.

File: tests/lineardev_tests.rs (hunk 11 above)
Comment (trgill) [inline tests/lineardev_tests.rs]: I'm thinking that the panic! macro is best avoided when there is useful cleanup that could be done.

In this case, if we panic  - the code that calls teardown() doesn't get executed.  So, the user of the test needs to use dmsetup to remove dangling DM devices.

File: src/engine/strat_engine/thindev.rs (hunk 7 above)
Comment (mulkieran) [inline src/engine/strat_engine/thindev.rs]: What about ```let dev_path = try!(self.path()).as_path();```.

If that doesn't work, might as well give up for now.

File: tests/lineardev_tests.rs (hunk 11 above)
Comment (mulkieran) [inline tests/lineardev_tests.rs]: ok.

File: tests/util/blockdev_utils.rs (hunk 10 above)
Comment (mulkieran) [inline tests/util/blockdev_utils.rs]: At some point, if we get enough of these, I might talk to Josh Stone. What is it doing?!!!

File: src/engine/strat_engine/thindev.rs (hunk 7 above)
Comment (trgill) [inline src/engine/strat_engine/thindev.rs]: :-)  I get the same error with that line.

Comment (mulkieran) [inline src/engine/strat_engine/thindev.rs]: sigh.

**Pull request merged**